# scrap-pangolin
<Introduction>
I am reruning data from the paper "Genomic analyses reveal poaching hotspots and illegal trade in pangolins from Africa to Asia." 
The specific pangolin from which the data comes from the White-bellied pangolin which is the most widespread species of pangolin in the African continent. 
They are also known as the Tree pangolins and are widely illegally traded. 
Being a vcf file and from the methods it has been quality checked but I was not completely a 100% sure that the SNPs and Indels were filtered for. 
This is why I modified the Snakefile workflow from class to rerun the data through the filter to check if this was indeed the case since 1 of my goals have been to filter data that they have provided. 
The files that I have for this are listed under the results directory under the rerun_snps_and_indels_merge. 


<Methods> 




